pkg/runtime/serializer/streaming/streaming.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package streaming implements encoder and decoder for streams
+// of runtime.Objects over io.Writer/Readers.
+package streaming
+
+import (
+	"bytes"
+	"io"
+
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+type Framer interface {
+	NewFrameReader(r io.Reader) io.Reader
+	NewFrameWriter(w io.Writer) io.Writer
+}
+
+type Encoder interface {
+	Encode(obj runtime.Object, overrides ...unversioned.GroupVersion) error
+}
+
+type Decoder interface {
+	Decode(defaults *unversioned.GroupVersionKind, into runtime.Object) (runtime.Object, *unversioned.GroupVersionKind, error)
+}
+
+type Serializer interface {
+	NewEncoder(w io.Writer) Encoder
+	NewDecoder(r io.Reader) Decoder
+}
+
+type decoder struct {
+	reader  io.Reader
+	decoder runtime.Decoder
+	buf     []byte
+}
+
+func NewDecoder(r io.Reader, d runtime.Decoder) Decoder {
+	return &decoder{
+		reader:  r,
+		decoder: d,
+		buf:     make([]byte, 1024*1024),
```

Comment:
Added a lot more comments.  I think this is point in time until the watch
PR is ready - that PR will make it a lot clearer what the code looks like
when we need to get a streaming codec for "application/json" that returns
newline delimited JSON blobs, but a length prefix framed binary protobuf
runtime.Unknown wrapping a raw protobuf byte array for
"application/vnd.protobuf".

On Wed, Mar 23, 2016 at 3:00 AM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In pkg/runtime/serializer/streaming/streaming.go
> <https://github.com/kubernetes/kubernetes/pull/23288#discussion_r57116863>
> :
>
> > +type Serializer interface {
> > +	NewEncoder(w io.Writer) Encoder
> > +	NewDecoder(r io.Reader) Decoder
> > +}
> > +
> > +type decoder struct {
> > +	reader  io.Reader
> > +	decoder runtime.Decoder
> > +	buf     []byte
> > +}
> > +
> > +func NewDecoder(r io.Reader, d runtime.Decoder) Decoder {
> > +	return &decoder{
> > +		reader:  r,
> > +		decoder: d,
> > +		buf:     make([]byte, 1024*1024),
>
> Yeah - SGTM. Can you please add a TODO?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23288/files/d1795d969a42d04fad220f4c75f2b95f1cfceb85#r57116863>
>